### PR TITLE
Integrate functional analyzers into the class-based analyzer pipeline

### PR DESCRIPTION
**Pull request by emallson**
Related: #4541

This replaces the current implementation of functional analyzers (which is a bespoke addition to the analysis pipeline) with an implementation that generates an anonymous class from a functional analyzer and then passes it through the existing class-based analyzer pipeline.

In the process, I also caught an issue in our integration test system that wasn't triggering the event sorting code that had been added to the functional section of the `CombatLogParser` when running integration tests and thus had to update some snapshots once this was moved to a location that *was* reached during integration tests. The code there was a hack in the first place and should've been done better but :shrug: it wasn't.

As I mentioned [here](https://github.com/WoWAnalyzer/WoWAnalyzer/issues/4541#issuecomment-1087762260), the goal is not to eliminate functional analyzers but to allow them to exist within the existing infrastructure without special-cased support in `CombatLogParser`. This allows us to use & experiment with them without having to worry about ramifications for the core analysis system.

There is still one piece of "special-cased" support: I added the ability to return `Suggestion[]` from the `suggestions()` method, which also applies to class analyzers but is currently not used by any. Could be useful if we want to strip away some of the magic around suggestions.

(The branch name is related to other stuff that I was planning to do at the same time, like eventdispatch -> normalizers, but this ended up being large enough that I wanted to PR it solo)

### CODE REVIEW

File: src/parser/core/Analyzer.ts [hunk 1]
```diff
@@ -50,3 +56,93 @@ class Analyzer extends EventSubscriber {
 }
 
 export default Analyzer;
+
+type ConstructedDependency<T> = T extends new (options: Options) => infer R ? R : never;
+
+type InjectedDependencies<Deps extends Dependencies> = {
+  [Key in keyof Deps]: ConstructedDependency<Deps[Key]>;
+};
```
Comment (emallson) [inline src/parser/core/Analyzer.ts:64]: also just wanna say: the typescript type system is pretty neat. there are still pretty severe limitations around dynamically generated class properties, but outside of those we can actually do neat things with dependencies.

if `Deps` is the dependency map you'd supply to `static dependencies`, then `InjectedDependencies<Deps>` is a map with each key being a constructed instance of the dependency. of limited type-strength right now due to the aforementioned issues with dynamic class props, but a small step is a step nonetheless.

File: src/parser/core/Analyzer.ts [hunk 2]
```diff
@@ -50,3 +56,93 @@ class Analyzer extends EventSubscriber {
...
+    eventList: AnyEvent[] = [];
+
+    constructor(options: Options) {
+      super(options);
+
+      eventFilters.forEach((filter) => this.addEventListener(filter, this.appendEvent));
+    }
+
+    private appendEvent(event: AnyEvent) {
+      this.eventList.push(event);
```
Comment (emallson) [inline src/parser/core/Analyzer.ts:97]: this does make a shallow copy of the event list, which is not ideal. I was lazy with the APL checker's `EventFilter` and left them at the default `EventFilter.any` even though they need something like `[EventFilter.cast, EventFilter.updatespellusable, EventFilter.begincast, ...]`. I need to find or write a function that applies an `EventFilter` to an array of events.

**Comment by MartijnHols** [comment]
I have a review pending, I just need to go over all my comments again to sanity check them when I am les tired. Will do that when it's not night time and I have time. If that takes too long just ping and I'll dump it as-is.

**Comment by emallson** [comment]
> I have a review pending, I just need to go over all my comments again to sanity check them when I am les tired. Will do that when it's not night time and I have time. If that takes too long just ping and I'll dump it as-is.

no rush, this isn't pressing.

**Pull request merged**